web/oauth.go

```diff
@@ -274,38 +274,49 @@ func completeOAuth(c *Context, w http.ResponseWriter, r *http.Request) {
 	body, teamId, props, tokenUser, err := c.App.AuthorizeOAuthUser(w, r, service, code, state, uri)
 
 	action := ""
+	hasRedirectURL := false
+	isActionMobileAuth := false
+	redirectURL := ""
 	if props != nil {
 		action = props["action"]
+		isActionMobileAuth = action == model.OAUTH_ACTION_MOBILE
+		if val, ok := props["redirect_to"]; ok {
+			redirectURL = val
+			hasRedirectURL = redirectURL != ""
+		}
 	}
 
-	if err != nil {
-		err.Translate(c.App.T)
-		mlog.Error(err.Error())
-		if action == model.OAUTH_ACTION_MOBILE {
-			w.Write([]byte(err.ToJson()))
+	renderError := func(err *model.AppError) {
+		if isActionMobileAuth {
+			if hasRedirectURL {
+				utils.RenderMobileError(c.App.Config(), w, err)
+			} else {
+				w.Write([]byte(err.ToJson()))
+			}
 		} else {
 			utils.RenderWebAppError(c.App.Config(), w, r, err, c.App.AsymmetricSigningKey())
 		}
+	}
+
+	if err != nil {
+		err.Translate(c.App.T)
+		mlog.Error(err.Error())
+		renderError(err)
 		return
 	}
 
 	user, err := c.App.CompleteOAuth(service, body, teamId, props, tokenUser)
 	if err != nil {
 		err.Translate(c.App.T)
 		mlog.Error(err.Error())
-		if action == model.OAUTH_ACTION_MOBILE {
-			w.Write([]byte(err.ToJson()))
-		} else {
-			utils.RenderWebAppError(c.App.Config(), w, r, err, c.App.AsymmetricSigningKey())
-		}
+		renderError(err)
 		return
 	}
 
-	var redirectUrl string
 	if action == model.OAUTH_ACTION_EMAIL_TO_SSO {
-		redirectUrl = c.GetSiteURLHeader() + "/login?extra=signin_change"
+		redirectURL = c.GetSiteURLHeader() + "/login?extra=signin_change"
 	} else if action == model.OAUTH_ACTION_SSO_TO_EMAIL {
-		redirectUrl = app.GetProtocol(r) + "://" + r.Host + "/claim?email=" + url.QueryEscape(props["email"])
+		redirectURL = app.GetProtocol(r) + "://" + r.Host + "/claim?email=" + url.QueryEscape(props["email"])
 	} else {
 		isMobile, parseErr := strconv.ParseBool(props[model.USER_AUTH_SERVICE_IS_MOBILE])
```

Comment:
Are `isMobile` and `isActionMobileAuth` always the same, should we be using one or the other?

Reply:
Yes, they both are the same. I have removed it from this function.